### PR TITLE
360. Sort Transformed Array

### PR DESCRIPTION
**Pull request by altayhunoglu**
(Refer to https://github.com/spiralgo/algorithms/issues/52 for detailed explanation)

It is possible to solve this question using PriorityQueue or built-in sort functions. 
Although I add the naive solution with PriorityQueue; it would be great to have some mathematical sense on the occasion of this great question:

A quadratic function of the form _**f(x) = ax2 + bx + c**_ is given.
The graph of a quadratic function called a **parabola**.

We have two possible cases (and one more insignificant case a == 0 as it will not affect the outcome.)
parabola

**Case 1:**
If a > 0 (positive) then the parabola opens upward.
 In this case, the middle "y" will be the minimum and one of the two ends (y1, y2) will be the maximum. 
 As y1, y2 must be greater than others; we need to compare these y pairs at each iteration and insert them into the end of the array.

 **Case 2:**
If a < 0 (negative) then the parabola opens downward.

 In this case, the middle "y" will be the maximum and one of the two ends (y1, y2) will be the minimum. 
 As y1, y2 must be smaller than others; we need to compare these y pairs at each iteration and insert them into the start of the array.

image

### CODE REVIEW

**Comment by ErdemT09** [comment]
My idea was the reduce to reduce unnecessary parameters and reappearing code. I think this won't make any performance any performance differences but this looks more modular to me.
```
public class Solution {

        int n, start = 0, end;
        int[] res = new int[n];
        int[] nums;
        int a, b, c;

        public int[] sortTransformedArray(int[] nums, int a, int b, int c) {
            n = nums.length;
            end = n-1;
            this.nums = nums;
            this.a = a;
            this.b = b;
            this.c = c;
            
            if (a > 0) {
                for (int i = n - 1; i >= 0; i--) {
                    quarticComparator(i, true);
                }
            } else {
                for (int i = 0; i < n; i++) {
                    quarticComparator(i, false);
                }
            }

            return res;
        }

        void quarticComparator(int i, boolean pbType) {
            int y1 = quadratic(nums[start]);
            int y2 = quadratic(nums[end]);
            if ((pbType && y1 > y2) || (!pbType && y1 < y2)) {
                res[i] = y1;
                start++;
            } else {
                res[i] = y2;
                end--;
            }
        }

        int quadratic(int x) {
            return a * x * x + b * x + c;
        }
    }
```

**Comment by altayhunoglu** [comment]
@ErdemT09 your refactoring is quite fine and reasonable.
It is why I'll be pushing back.

image

File: src/main/java/algorithms/curated170/medium/SortTransformedArray.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            res = new int[n];
+            end = n-1;
+            this.nums = nums;
+            this.a = a;
+            this.b = b;
+            this.c = c;
+            
+            if (a > 0) {
+                for (int i = n - 1; i >= 0; i--) {
+                    quarticComparator(i, true);
```
Comment (ErdemT09) [inline src/main/java/algorithms/curated170/medium/SortTransformedArray.java:21]: This should be called quadraticComparator, quartic is something relating to the power of 4 (x^4).

**Review by ErdemT09 — APPROVED** [review]
This question is another example of a two pointers problem. It also shows how mathematics can come up as important in algorithms. I give my approval.

**Pull request merged**